Classes/Controller/TableOfContentsController.php

```diff
@@ -33,6 +33,167 @@ class TableOfContentsController extends AbstractController
+                    && $this->requestData['page'] < $this->document->getDoc()->numPages
+                ) {
+                    $this->activeEntries = array_merge($this->activeEntries,
+                        (array) $this->document->getDoc()->smLinks['p2l'][$this->document->getDoc()->physicalStructure[$this->requestData['page'] + 1]]);
+                }
+            }
+            // Go through table of contents and create all menu entries.
+            foreach ($this->document->getDoc()->tableOfContents as $entry) {
+                $menuArray[] = $this->getMenuEntry($entry, true);
+            }
+        } else {
+            // Go through table of contents and create top-level menu entries.
+            foreach ($this->document->getDoc()->tableOfContents as $entry) {
+                $menuArray[] = $this->getMenuEntry($entry, false);
+            }
+            // Build table of contents from database.
+            $result = $this->documentRepository->getTableOfContentsFromDb($this->document->getUid(), $this->document->getPid(), $this->settings);
+
+            $allResults = $result->fetchAll();
+
+            if (count($allResults) > 0) {
+                $menuArray[0]['ITEM_STATE'] = 'CURIFSUB';
+                $menuArray[0]['_SUB_MENU'] = [];
+                foreach ($allResults as $resArray) {
+                    $entry = [
+                        'label' => !empty($resArray['mets_label']) ? $resArray['mets_label'] : $resArray['title'],
+                        'type' => $resArray['type'],
+                        'volume' => $resArray['volume'],
+                        'orderlabel' => $resArray['mets_orderlabel'],
+                        'pagination' => '',
+                        'targetUid' => $resArray['uid']
+                    ];
+                    $menuArray[0]['_SUB_MENU'][] = $this->getMenuEntry($entry, false);
+                }
+            }
+        }
+        return $menuArray;
+    }
+
+    /**
+     * This builds a menu for list of 3D objects
+     *
+     * @access protected
+     *
+     * @param string $content: The PlugIn content
+     * @param array $conf: The PlugIn configuration
+     *
+     * @return array HMENU array
+     */
+    protected function makeMenuFor3DObjects()
+    {
+        // Set default values for page if not set.
+        // $this->requestData['page'] may be integer or string (physical structure @ID)
+        if (
+            (int) $this->requestData['page'] > 0
+            || empty($this->requestData['page'])
+        ) {
+            $this->requestData['page'] = MathUtility::forceIntegerInRange((int) $this->requestData['page'], 1, $this->document->getDoc()->numPages, 1);
+        } else {
+            $this->requestData['page'] = array_search($this->requestData['page'], $this->document->getDoc()->physicalStructure);
+        }
+        $this->requestData['double'] = MathUtility::forceIntegerInRange($this->requestData['double'], 0, 1, 0);
```

Comment:
Is this really necessary? When dealing with an 3D object list there are no physical pages...

Reply:
Removed